validator/validators/performance/performance_validator.go

```diff
@@ -57,3 +90,153 @@ func (s *PerformanceValidator) Cleanup() error {
+		log.Printf("Start calculate metric statictics for metric %s %v", metricName, metricValues)
+		if !isAllValuesGreaterThanOrEqualToZero(metricValues) {
+			return nil, fmt.Errorf("values are not all greater than or equal to zero for metric %s with values: %v", metricName, metricValues)
+		}
+		metricStats := CalculateMetricStatisticsBasedOnDataAndPeriod(metricValues, agentCollectionPeriod)
+		log.Printf("Finished calculate metric statictics for metric %s: %v", metricName, metricStats)
+		performanceMetricResults[metricName] = metricStats
+	}
+
+	return packIntoPerformanceInformation(receiver, dataType, fmt.Sprint(agentCollectionPeriod), commitHash, commitDate, map[string]interface{}{dataRate: performanceMetricResults}), nil
+}
+
+func (s *PerformanceValidator) GetPerformanceMetrics(startTime, endTime time.Time) ([]types.MetricDataResult, error) {
+	var (
+		metricNamespace              = s.vConfig.GetMetricNamespace()
+		validationMetric             = s.vConfig.GetMetricValidation()
+		ec2InstanceId                = awsservice.GetInstanceId()
+		performanceMetricDataQueries = []types.MetricDataQuery{}
+	)
+	log.Printf("Start getting performance metrics from CloudWatch")
+	for _, metric := range validationMetric {
+		metricDimensions := []types.Dimension{
+			{
+				Name:  aws.String("InstanceId"),
+				Value: aws.String(ec2InstanceId),
+			},
+		}
+		for _, dimension := range metric.MetricDimension {
+			metricDimensions = append(metricDimensions, types.Dimension{
+				Name:  aws.String(dimension.Name),
+				Value: aws.String(dimension.Value),
+			})
+		}
+		performanceMetricDataQueries = append(performanceMetricDataQueries, s.buildStressMetricQueries(metric.MetricName, metricNamespace, metricDimensions))
+	}
+
+	metrics, err := awsservice.GetMetricData(performanceMetricDataQueries, startTime, endTime)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return metrics.MetricDataResults, nil
+}
+
+func (s *PerformanceValidator) buildStressMetricQueries(metricName, metricNamespace string, metricDimensions []types.Dimension) types.MetricDataQuery {
+	metricInformation := types.Metric{
+		Namespace:  aws.String(metricNamespace),
+		MetricName: aws.String(metricName),
+		Dimensions: metricDimensions,
+	}
+
+	metricDataQuery := types.MetricDataQuery{
+		MetricStat: &types.MetricStat{
+			Metric: &metricInformation,
+			Period: aws.Int32(10),
+			Stat:   aws.String(string(models.AVERAGE)),
+		},
+		Id: aws.String(strings.ToLower(metricName)),
+	}
+	return metricDataQuery
+}
+
+// packIntoPerformanceInformation will package all the information into the required format of MongoDb Database
+// https://github.com/aws/amazon-cloudwatch-agent-test/blob/e07fe7adb1b1d75244d8984507d3f83a7237c3d3/terraform/setup/main.tf#L8-L63
+func packIntoPerformanceInformation(receiver, dataType, collectionPeriod, commitHash string, commitDate int64, result interface{}) PerformanceInformation {
+	instanceAMI := awsservice.GetImageId()
+	instanceType := awsservice.GetInstanceType()
```

Comment:
Any chance GetImageId() or GetInstanceType could fail?
If so, what can we do to handle that type of error?
1. log error?
2. retry?
3. fail the test?

Reply:
```
Any chance GetImageId() or GetInstanceType could fail?
```
It would be less likely only in the case of turning off IMDS. However, this could be prevented by the following options:
* Instead of calling IMDS here, we could pass it as an parameters in the [generator configuration](https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/validator/models/validation_config.go#L29-L48) (e.g instance_id in parameters.yml)  since [we know this kind of information when we create the EC2 instances](https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/terraform/performance/main.tf#L54-L69) and get it there

Reply:
I am fine leaving it as is.
The error seems very unlikely to happen as long as we control the instance the test runs on and properly turn on IMDS. I think logging a failure is good enough. Up to you though.